Code/Tools/AssetProcessor/Platform/Linux/native/FileWatcher/FileWatcher_linux.cpp

```diff
@@ -34,38 +71,70 @@ bool FileWatcher::PlatformImplementation::Initialize()
         [[maybe_unused]] AZStd::fixed_string<255> errorString;
         AZ_Warning("FileWatcher", (m_inotifyHandle >= 0), "Unable to initialize inotify, file monitoring will not be available: %s\n", strerror_r(err, errorString.data(), errorString.capacity()));
     }
+
+    // create a handle that you can write to, in order to wake up the listening thread immediately.
+    m_wakeThreadHandle = eventfd(0, EFD_CLOEXEC | EFD_SEMAPHORE);
+
     return (m_inotifyHandle >= 0);
 }
 
-void FileWatcher::PlatformImplementation::Finalize()
+void FileWatcher::PlatformImplementation::CloseMainWatchHandle()
 {
     if (m_inotifyHandle < 0)
     {
         return;
     }
+      
+    ::close(m_inotifyHandle);
+    m_inotifyHandle = -1;
 
+    // signal the thread to awaken in case its blocked on read.
+    AZ::u64 flagValue = 1;
+    if (write(m_wakeThreadHandle, &flagValue, sizeof(flagValue)) != sizeof(flagValue))
     {
-        QMutexLocker lock{&m_handleToFolderMapLock};
-        for (const auto& watchHandle : m_handleToFolderMap.keys())
-        {
-            inotify_rm_watch(m_inotifyHandle, watchHandle);
-        }
-        m_handleToFolderMap.clear();
+        AZ_ErrorOnce("FileWatcher", false, "Was unable to write to the wake-up-the-thread handle, this may cause timeouts and deadlocks.\n");
     }
+}
 
-    ::close(m_inotifyHandle);
-    m_inotifyHandle = -1;
+void FileWatcher::PlatformImplementation::Finalize()
+{
+    // This happens after the watch thread has already joined.
+    
+    // note from the man-page of inotify " all associated watches are automatically freed." - no need to call inotify_rm_watch on each handle.
+    if (m_wakeThreadHandle >= 0)
+    {
+        close(m_wakeThreadHandle);
+    }
```

Comment:
For the safety should we reset m_wakeThreadHandle back to `-1`?

Reply:
I'll add that.